utilities/xmlrpcpp/src/XmlRpcClient.cpp

```diff
@@ -322,6 +347,8 @@ XmlRpcClient::writeRequest()
   // Try to write the request
   if ( ! XmlRpcSocket::nbWrite(this->getfd(), _request, &_bytesWritten)) {
     XmlRpcUtil::error("Error in XmlRpcClient::writeRequest: write error (%s).",XmlRpcSocket::getErrorMsg().c_str());
+    // If the write fails, we had an unrecoverable error. Close the socket.
+    close();
```

Comment:
Instead of calling `close()` in the various locations where a read / write fails would it be possible that the calling code actually does that when receiving `false` instead? Why should this be added here and change the behavior of the API?

Reply:
No; there is already a contract between this code and the calling code that determines if the caller should call close on error or not, based on the return value of the setKeepOpen() function: https://github.com/ros/ros_comm/blob/8b741ac9232a9fadefb19e7fdee654225619d2f8/utilities/xmlrpcpp/src/XmlRpcDispatch.cpp#L157-L160

It looks like this is used in the client to keep the socket open after a request is complete so that the socket is available for subsequent requests. I didn't want to change the behavior of that mechanism or interfere with the concept of keeping a client TCP socket open between requests so I opted to keep the logic within the XmlRpcClient class.

I contemplated a few different ways to structure this change:
1. Make the client code call close in the correct scenarios, but this breaks encapsulation and further exposes the client to implementation details of this class.
1. Make the XmlRpcSocket class close the socket on error. This is cleaner, but would require a complete rewrite of XmlRpcSocket to make it own the file descriptor. This was a bigger set of changes than I wanted to make and I thought they would be seen as too invasive to be accepted into ros_comm.
1. Test XmlRpcSocket to make sure that it always returns false when there is an error, and then fix XmlRpcClient so that it always closes the socket when there is an error. I deemed this to be the least invasive and least likely to include changes that could inadvertently change the library behavior.

This is not a breaking change to the API; file descriptors are only closed in places where they were already in an error state, and it's safe to call close multiple times, so client code which calls close again will not fail. Any client which was expecting a persistent connection and which ends up with a broken connection will behave better than the previous implementation. In the previous implementation it would discover the broken socket when attempting to send a new request and would be forced to close the socket and reconnect. In the updated implementation the socket is already closed, so the client immediately knows that it needs to establish a new connection.